controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,407 @@
+		WithLabels(map[string]string{"foo": "bar"}).
+		WithAnnotations(map[string]string{"foo": "bar"}).
+		Build()
+	machineDeploymentClass2 := builder.MachineDeploymentClass(ns.Name, "md2").
+		WithClass(workerClassName2).
+		WithInfrastructureTemplate(infrastructureMachineTemplate).
+		WithBootstrapTemplate(bootstrapTemplate).
+		Build()
+	clusterClass := builder.ClusterClass(ns.Name, clusterClassName).
+		WithInfrastructureClusterTemplate(infrastructureClusterTemplate).
+		WithControlPlaneTemplate(controlPlaneTemplate).
+		WithControlPlaneInfrastructureMachineTemplate(infrastructureMachineTemplate).
+		WithWorkerMachineDeploymentClasses([]clusterv1.MachineDeploymentClass{*machineDeploymentClass1, *machineDeploymentClass2}).
+		Build()
+
+	// 3) A Cluster object including a Cluster Topology Object and the MachineDeploymentTopology objects used in the
+	// Cluster Topology.
+	machineDeploymentTopology1 := builder.MachineDeploymentTopology("mdm1").
+		WithClass(workerClassName1).
+		WithReplicas(3).
+		Build()
+	machineDeploymentTopology2 := builder.MachineDeploymentTopology("mdm2").
+		WithClass(workerClassName2).
+		WithReplicas(1).
+		Build()
+	cluster := builder.Cluster(ns.Name, clusterName).
+		WithTopology(
+			builder.ClusterTopology().
+				WithClass(clusterClass.Name).
+				WithMachineDeployment(machineDeploymentTopology1).
+				WithMachineDeployment(machineDeploymentTopology2).
+				WithVersion("1.22.2").
+				WithControlPlaneReplicas(3).
+				Build()).
+		Build()
+
+	// Create a set of initObjects from the objects above to add to the API server when the test environment starts.
+	initObjs := []client.Object{
+		clusterClass,
+		cluster,
+		infrastructureClusterTemplate,
+		infrastructureMachineTemplate,
+		bootstrapTemplate,
+		controlPlaneTemplate,
+	}
+
+	for _, obj := range initObjs {
+		g.Expect(env.Create(ctx, obj)).To(Succeed())
+	}
+	defer func() {
+		for _, obj := range initObjs {
+			g.Expect(env.Delete(ctx, obj)).To(Succeed())
+		}
+	}()
+
+	g.Eventually(func(g Gomega) error {
+		// Get the cluster object.
+		key := client.ObjectKey{Name: clusterName, Namespace: ns.Name}
+		actualCluster := &clusterv1.Cluster{}
+		if err := env.Get(ctx, key, actualCluster); err != nil {
+			return err
+		}
+
+		// Check if Cluster has relevant Infrastructure and ControlPlane and labels and annotations.
+		g.Expect(assertClusterReconcile(actualCluster)).Should(Succeed())
+
+		// Check if InfrastructureCluster has been created and has the correct labels and annotations.
+		g.Expect(assertInfrastructureClusterReconcile(actualCluster)).Should(Succeed())
+
+		// Check if ControlPlane has been created and has the correct version, replicas, labels and annotations.
+		g.Expect(assertControlPlaneReconcile(actualCluster, clusterClass)).Should(Succeed())
+
+		// Check if MachineDeployments are created and have the correct version, replicas, labels annotations and templates.
+		g.Expect(assertMachineDeploymentsReconcile(actualCluster)).Should(Succeed())
```

Comment:
```suggestion
		if err := assertClusterReconcile(actualCluster); err != nil {
			return err
		}

		// Check if InfrastructureCluster has been created and has the correct labels and annotations.
		if err := assertInfrastructureClusterReconcile(actualCluster); err != nil {
			return err
		}

		// Check if ControlPlane has been created and has the correct version, replicas, labels and annotations.
		if err := assertControlPlaneReconcile(actualCluster, clusterClass); err != nil {
			return err
		}

		// Check if MachineDeployments are created and have the correct version, replicas, labels annotations and templates.
		return assertMachineDeploymentsReconcile(actualCluster)
```

I think you have to implement it that way, otherwise the test fails if the first try doesn't succeed (as .Expect fails the test immediately).

Locally the test failed for me without this change

Reply:
Can you share how you're making the test fail? for me if I return an error in the first line of assertClusterReconcile it retries until timeout as expected.

Reply:
Interesting :). I just started it with debug, added a break point, continued and it failed. Let me re-test.

Reply:
My bad, the break point did trigger the timeout. So TIL g.Expect does not fail immediately in Eventually.